File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

// SendAll sends all bytes over the provided io.Writer, ensuring that all data is sent without short write errors.
func SendAll(socket io.Writer, bytes []byte) error {
	totalSent := 0

	for totalSent < len(bytes) {
		sent, err := socket.Write(bytes[totalSent:])

		if err != nil {
			return err
		}
		totalSent += sent
	}

	return nil
}

// RecvAll receives the specified number of bytes from the provided io.Reader, ensuring that all data is received
// without short read errors.
func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalRecv := 0

	for totalRecv < size {
		recv, err := socket.Read(buff[totalRecv:])
		totalRecv += recv

		if err != nil {
			return nil, err
		}
	}

	return buff, nil
}

File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(sock: socket.socket, size: int) -> bytes:
    """
    Receives a specified number of bytes from a socket, ensuring that the exact amount is received.
    """
    received = bytearray()
    while len(received) < size:
        bytes_read = sock.recv(size - len(received))
        if not bytes_read:
            break
        received.extend(bytes_read)
    return bytes(received)


def send_all(socket: socket.socket, bytes):
    """
    Sends all bytes to a socket, ensuring that the exact amount is sent.
    """
    total_sent = 0
    while total_sent < len(bytes):
        sent = socket.send(bytes[total_sent:])
        total_sent += sent
    return total_sent
